lib/internal/fs/dir.js

```diff
@@ -0,0 +1,179 @@
+'use strict';
+
+const { Object } = primordials;
+
+const pathModule = require('path');
+const binding = internalBinding('fs');
+const dirBinding = internalBinding('fs_dir');
+const {
+  codes: {
+    ERR_INVALID_CALLBACK,
+    ERR_MISSING_ARGS
+  }
+} = require('internal/errors');
+
+const { FSReqCallback } = binding;
+const internalUtil = require('internal/util');
+const {
+  getDirent,
+  getOptions,
+  getValidatedPath,
+  handleErrorFromBinding
+} = require('internal/fs/utils');
+
+const kDirHandle = Symbol('kDirHandle');
+const kDirPath = Symbol('kDirPath');
+const kDirOptions = Symbol('kDirOptions');
+const kDirReadPromisified = Symbol('kDirReadPromisified');
+const kDirClosePromisified = Symbol('kDirClosePromisified');
+
+class Dir {
+  constructor(handle, path, options) {
+    if (handle == null) throw new ERR_MISSING_ARGS('handle');
+    this[kDirHandle] = handle;
+    this[kDirPath] = path;
+
+    this[kDirOptions] = getOptions(options, {
+      encoding: 'utf8'
+    });
+
+    this[kDirReadPromisified] = internalUtil.promisify(this.read).bind(this);
+    this[kDirClosePromisified] = internalUtil.promisify(this.close).bind(this);
+  }
+
+  get path() {
+    return pathModule.resolve(this[kDirPath]);
```

Comment:
Another thought now that I’m reading this again … if the cwd or the directory structure changes between creating the file handle and this `path.resolve()` call, the resulting path may be invalid.

I think it might be best to not call `path.resolve()` on it at all, and just return what the user passed in at the beginning? That would be easiest, leaves responsibility for handling the result correctly to the user, and it works fine even when `path` is a `Buffer` (+ we should probably add a test for that last one, too).

Reply:
Ideally I'd like to change/add stuff to libuv so this goes and asks libuv - is that reasonable?

_(It should probably be an actual function then though.)_

That being said, it doesn't actually seem like there is a way to get a file path from fd either, so maybe it's unimportant. It would be nice to have something for debugging purposes though.

Perhaps we should just give it a fd like would exist on unix and emulatable on windows?

Reply:
> That being said, it doesn't actually seem like there is a way to get a file path from fd either, so maybe it's unimportant.

Yes, there is no reliable way to get a path from an fd, in part because there may not be a corresponding valid path for an fd anymore or a fd can refer to multiple paths (although for directories hard links are generally disallowed).

I don’t think libuv can generally do something better here, either. And for debugging, the original value should be fine, I would guess…